### PR TITLE
#164: bug with multiple registrations by changing case fixed

### LINKED ISSUE #164: You can re-register your mail by changing case
You can register multiple accounts by changing case in email. For example: email@gmail.com and Email@gmail.com perceived as a different mail in database.

### PR DESCRIPTION
**Pull request by IgorMaksymov**
#164: bug with multiple registrations by changing case fixed
connect to #164

### CODE REVIEW

**Comment by IgorMaksymov** [comment]
@AndriyBaibak time spent: 30m

**Comment by vaa25** [comment]
@IgorMaksymov no tests

**Comment by IgorMaksymov** [comment]
@vaa25 test for multiple registrations is not working.
issue https://github.com/JuniorsJava/itevents/issues/155

**Comment by IgorMaksymov** [comment]
blocked with #155 @AndriyBaibak @romach

**Comment by IgorMaksymov** [comment]
@vaa25 its not blocked anymore

**Comment by vaa25** [comment]
@IgorMaksymov no test showing solved problem in task

**Comment by vaa25** [comment]
@AndriyBaibak 50 min with other reviews today before

**Comment by IgorMaksymov** [comment]
@vaa25 fixed

**Comment by vaa25** [comment]
@IgorMaksymov It must be integration test. it shoud try register new user via controller through service and dao. Dao have to work with dbunit. dbunit must contain similar login. Test have to check response code and message.
Also last commit has conflicts with master

**Comment by IgorMaksymov** [comment]
@vaa25 my unit test shows that in DAO layer user comes with LOWERCASE, but in SERVICE layer incoming parameter is in UPPERCASE.
we dont need such complicated test, as you described.

**Comment by IgorMaksymov** [comment]
@vaa25 fixed

**Comment by romach** [comment]
@IgorMaksymov 
I think that manipulations with case of username in `UserService` is a bad idea. It is better to use validation in `UserRestController.registerNewSubscriber()` to solve this issue.

**Comment by romach** [comment]
@AndriyBaibak time spent: 30 minutes

**Comment by IgorMaksymov** [comment]
@romach you want @email validation? 
the problem is, that in database saves "emAil", when there already exist login "email"

or you want to send message to user, that he must type his login with lovercase?
in this case, its better to handle this in our side.

**Comment by vaa25** [comment]
@romach are this manipulations are part of business logic?

**Comment by romach** [comment]
> are this manipulations are part of business logic?

@vaa25 yes, they are. But I think that this business logic is wrong. For example, I try to register with username `Roma@mail.com` (such username is absent in DB). [`user.setLogin(user.getLogin().toLowerCase());`](https://github.com/JuniorsJava/itevents/commit/2400edc36c3ef66497457b6331688b4a7efe00c4#diff-1a44d5d5bc0e25b38653c34e684d897cR55) changes my username and do not throw any errors that this username is wrong. After registration I will figure out that my username is `roma@mail.com` (maybe I don't want this). I think that:
1) if uppercase letters in username are not allowed, exception must be throwed to user;
2) our system shouldn't change user input data;

**Comment by IgorMaksymov** [comment]
@romach 
1) login is an EMAIL, so it should be in lowercase.
2) bothering user with exceptions, because his email not like we need, is not good, so i decided to handle this in our side.
3) i dont see possibilities to make 'Roma/romA/roma/rOMa' to be the same for our database.

**Comment by romach** [comment]
@IgorMaksymov 
> 1) login is an EMAIL, so it should be in lowercase.

according to [Wikipedia](https://en.wikipedia.org/wiki/Email_address#Local_part) email can contain uppercase letters;
> 3) i dont see possibilities to make 'Roma/romA/roma/rOMa' to be the same for our database.

ok, maybe it is better to create separate method `checkEmailForUniqIgnoreCase()` in `UserService` to which we will pass lowercased username but in database we will store original username.

**Comment by IgorMaksymov** [comment]
@romach yes, it can, and its 2 different emails, read further.
this is overengineering, we have simple .toLowerCase, other will not make code more readable..

**Comment by romach** [comment]
@IgorMaksymov resolve merge conflicts, please

**Comment by IgorMaksymov** [comment]
@romach done

**Comment by vaa25** [comment]
@IgorMaksymov where is integration test?

**Comment by IgorMaksymov** [comment]
@vaa25 as is said before, we dont need it.
All manipulations is handled in current tests.
if you remove .toLoverCase() test will fail.

**Comment by alex-anakin** [comment]
@IgorMaksymov 
1. Issue describe problem: **You can re-register your mail by changing case**. So your solution should forbid register user with same email in different cases. I didn't find test that shows how user tries to register with the same email in different cases and fails.

2. For example user registered with login *Address@mail.com*. Can user log in with the same login?

3. You make adding the new user with login forcibly formatted to lowercase. Why do you format username to lower case when getting it from the database?

**Comment by IgorMaksymov** [comment]
@alex-anakin 
1) yes, he can
2) database will not find user  Address@mail.com even if  address@mail.com exist

**Comment by vaa25** [comment]
@IgorMaksymov If user has Address@mail.com, but we store address@mail.com, will otp reach right destination?

**Comment by IgorMaksymov** [comment]
@vaa25 we can move our toLoverCase() to addSubscriver(), also i sent email from our service to adress IVMaksymov@gmail.com and received it.

**Comment by IgorMaksymov** [comment]
@romach passed or not?

**Comment by romach** [comment]
@IgorMaksymov issue branch builds with exception:
```
org.itevents.integration.UserRestControllerIntegrationTest > shouldExpect422IfRegisteringExistingLoginWithUpperCase FAILED
    java.lang.IllegalStateException: Failed to load ApplicationContext
                                                        
        Caused by:                                      
        org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'sendgrid' defined in class path resource [ap
plicationContextIntegrationTestAddon.xml]: Could not resolve placeholder 'sendgrid.api.key' in string value "${sendgrid.api.key}"; nested exception is
java.lang.IllegalArgumentException: Could not resolve placeholder 'sendgrid.api.key' in string value "${sendgrid.api.key}"
                                                        
            Caused by:                                  
            java.lang.IllegalArgumentException: Could not resolve placeholder 'sendgrid.api.key' in string value "${sendgrid.api.key}"
```

**Comment by IgorMaksymov** [comment]
@romach TRAVIS build works fine.
update properties file, or i forgot to add properties to a template, please heck this.
Take properties from Travis properties

**Comment by romach** [comment]
@IgorMaksymov 
>@romach TRAVIS build works fine.
update properties file, or i forgot to add properties to a template, please check this.
Take properties from Travis properties
i've made some additional properties file.

Fix `test-local.properties.TEMPLATE` because we must give properties to `local.properties` from it

File: restservice/src/test/java/org/itevents/service/transactional/MyBatisUserServiceTest.java [hunk 1]
```diff
@@ -313,4 +315,22 @@ public void shouldThrowServiceExceptionIfOtpIsInvalid() throws Exception {
 
         userService.activateUserWithOtp(stringOtp);
     }
+
+    @Test
+    public void shouldAddLoginInLowerCase() throws Exception {
```
Comment (romach) [inline restservice/src/test/java/org/itevents/service/transactional/MyBatisUserServiceTest.java]: @IgorMaksymov `shouldAddSubscriberWithLowerCasedLogin` is more understandable method name

**Comment by IgorMaksymov** [comment]
@romach i dont think that this is more readable.
fixed previous mistakes.

**Comment by vaa25** [comment]
@IgorMaksymov you shoudn't create own applicationContext for test

**Comment by IgorMaksymov** [comment]
@vaa25 i cannot make dynamic properties insertion or  refactor our properties within this task.
i propose add new task of removing excess config files, for not to slow down our development with blockers.

**Comment by vaa25** [comment]
@IgorMaksymov  remove duplicated variables from local.properties.TEMPLATE (look global.properties)
@AndriyBaibak 10min

**Comment by IgorMaksymov** [comment]
@vaa25 it should be done within #221

**Comment by vaa25** [comment]
@IgorMaksymov it added in this pr

File: restservice/src/test/resources/test-local.properties.TEMPLATE [hunk 2]
```diff
@@ -12,10 +12,6 @@ changelog.file=src\\test\\resources\\liquibase\\db.changelog-create.xml
 # SendGrid
 sendgrid.api.key=SEND_GRID_TOKEN
 
-# EventService
-event.filter.range.in.days=7
```
Comment (vaa25) [inline restservice/src/test/resources/test-local.properties.TEMPLATE]: @IgorMaksymov what will be with tests if somebody changes default variables in future?

Comment (IgorMaksymov) [inline restservice/src/test/resources/test-local.properties.TEMPLATE]: @vaa25 removed line was unused.

**Pull request merged**